File: examples_android/app/src/main/java/com/confirmit/testsdkapp/layouts/defaultlayoutjava/items/DefaultLayoutItemFactory.java
package com.confirmit.testsdkapp.layouts.defaultlayoutjava.items;

import com.confirmit.mobilesdk.ui.SurveyPage;
import com.confirmit.mobilesdk.ui.questions.DefaultQuestion;
import com.confirmit.mobilesdk.ui.questions.InfoQuestion;
import com.confirmit.mobilesdk.ui.questions.MultiQuestion;
import com.confirmit.mobilesdk.ui.questions.NumericQuestion;
import com.confirmit.mobilesdk.ui.questions.QuestionAnswer;
import com.confirmit.mobilesdk.ui.questions.QuestionText;
import com.confirmit.mobilesdk.ui.questions.SingleQuestion;
import com.confirmit.mobilesdk.ui.questions.TextQuestion;
import com.confirmit.testsdkapp.R;

import java.util.ArrayList;
import java.util.List;

class DefaultLayoutItemFactory {
    private static List<DefaultLayoutItem> buildCommon(DefaultQuestion question) {
        List<DefaultLayoutItem> result = new ArrayList<>();
        if (!question.getTitle().isEmpty()) {
            result.add(new DefaultLayoutIconLabel(question.getTitle(), R.drawable.ic_assignment_black));
        }
        if (!question.getText().isEmpty()) {
            result.add(new DefaultLayoutLabel(DefaultLabelType.TEXT, question.getText()));
        }
        if (!question.getInstruction().isEmpty()) {
            result.add(new DefaultLayoutLabel(DefaultLabelType.INSTRUCTION, question.getInstruction()));
        }
        if (!question.getErrors().isEmpty()) {
            result.add(new DefaultLayoutIconLabel(question.getErrors()));
        }

        return result;
    }

    static List<DefaultLayoutItem> createPageHeader(SurveyPage page) {
        List<DefaultLayoutItem> result = new ArrayList<>();
        if (!page.getTitle().isEmpty()) {
            result.add(new DefaultLayoutLabel(DefaultLabelType.PAGE_TITLE, page.getTitle()));
        }

        return result;
    }

    static List<DefaultLayoutItem> createInfo(InfoQuestion question) {
        List<DefaultLayoutItem> result = new ArrayList<>();
        if (!question.getTitle().isEmpty()) {
            result.add(new DefaultLayoutIconLabel(question.getTitle(), R.drawable.ic_assignment_black));
        }

        if (!question.getText().isEmpty()) {
            result.add(new DefaultLayoutLabel(DefaultLabelType.TEXT, question.getText()));
        }

        if (!question.getInstruction().isEmpty()) {
            result.add(new DefaultLayoutLabel(DefaultLabelType.INSTRUCTION, question.getInstruction()));
        }
        return result;
    }

    static List<DefaultLayoutItem> createText(TextQuestion question) {
        List<DefaultLayoutItem> result = buildCommon(question);
        result.add(new DefaultLayoutTextBox(DefaultLayoutTextBoxType.TEXT, question.getValue()));
        return result;
    }

    static List<DefaultLayoutItem> createNumeric(NumericQuestion question) {
        List<DefaultLayoutItem> result = buildCommon(question);
        result.add(new DefaultLayoutTextBox(DefaultLayoutTextBoxType.NUMERIC, question.getValue()));
        return result;
    }

    static List<DefaultLayoutItem> createNotSupported() {
        List<DefaultLayoutItem> result = new ArrayList<>();
        result.add(new DefaultLayoutIconLabel(new QuestionText("Question is not supported", false), R.drawable.ic_info_black));
        return result;
    }

    static List<DefaultLayoutItem> createSingle(SingleQuestion question) {
        List<DefaultLayoutItem> result = buildCommon(question);

        String selectedCode = "";
        if (question.selected() != null) {
            selectedCode = question.selected().getCode();
        }

        for (QuestionAnswer answer : question.getAnswers()) {
            if (answer.isHeader()) {
                if (!answer.getText().isEmpty()) {
                    result.add(new DefaultLayoutLabel(DefaultLabelType.TEXT, answer.getText(), 1));
                }
            } else {
                result.add(new DefaultLayoutClickLabel(DefaultLayoutClickLabelType.RADIO, selectedCode == answer.getCode(), 1, answer));
            }

            for (QuestionAnswer nestedAnswer : answer.getAnswers()) {
                result.add(new DefaultLayoutClickLabel(DefaultLayoutClickLabelType.RADIO, selectedCode == nestedAnswer.getCode(), 2, nestedAnswer));
            }
        }

        return result;
    }

    static List<DefaultLayoutItem> createMulti(MultiQuestion question) {
        List<DefaultLayoutItem> result = buildCommon(question);

        for (QuestionAnswer answer : question.getAnswers()) {
            if (answer.isHeader()) {
                if (!answer.getText().isEmpty()) {
                    result.add(new DefaultLayoutLabel(DefaultLabelType.TEXT, answer.getText(), 1));
                }
            } else {
                result.add(new DefaultLayoutClickLabel(DefaultLayoutClickLabelType.CHECKBOX, question.get(answer), 1, answer));
            }

            for (QuestionAnswer nestedAnswer : answer.getAnswers()) {
                result.add(new DefaultLayoutClickLabel(DefaultLayoutClickLabelType.CHECKBOX, question.get(nestedAnswer), 2, nestedAnswer));
            }
        }

        return result;
    }
}

File: examples_android/app/src/main/java/com/example/myapp/layouts/defaultlayout/items/DefaultLayoutItemFactory.kt
package com.example.myapp.layouts.defaultlayout.items

import com.confirmit.mobilesdk.ui.SurveyPage
import com.confirmit.mobilesdk.ui.questions.*
import com.example.myapp.R

object DefaultLayoutItemFactory {
    private fun buildCommon(question: DefaultQuestion): MutableList<DefaultLayoutItem> {
        val result = mutableListOf<DefaultLayoutItem>()
        if (!question.title.isEmpty()) {
            result.add(DefaultLayoutIconLabel(question.title, R.drawable.ic_assignment_black))
        }
        if (!question.text.isEmpty()) {
            result.add(DefaultLayoutLabel(DefaultLabelType.TEXT, question.text))
        }
        if (!question.instruction.isEmpty()) {
            result.add(DefaultLayoutLabel(DefaultLabelType.INSTRUCTION, question.instruction))
        }
        if (question.errors.isNotEmpty()) {
            result.add(DefaultLayoutIconLabel(question.errors))
        }

        return result
    }

    fun createPageHeader(page: SurveyPage): MutableList<DefaultLayoutItem> {
        val result = mutableListOf<DefaultLayoutItem>()
        if (!page.title.isEmpty()) {
            result.add(DefaultLayoutLabel(DefaultLabelType.PAGE_TITLE, page.title))
        }

        return result
    }

    fun createInfo(question: InfoQuestion): MutableList<DefaultLayoutItem> {
        val result = mutableListOf<DefaultLayoutItem>()
        if (!question.title.isEmpty()) {
            result.add(DefaultLayoutIconLabel(question.title, R.drawable.ic_assignment_black))
        }

        if (!question.text.isEmpty()) {
            result.add(DefaultLayoutLabel(DefaultLabelType.TEXT, question.text))
        }

        if (!question.instruction.isEmpty()) {
            result.add(DefaultLayoutLabel(DefaultLabelType.INSTRUCTION, question.instruction))
        }

        return result
    }

    fun createText(question: TextQuestion): MutableList<DefaultLayoutItem> {
        val result = buildCommon(question)
        result.add(DefaultLayoutTextBox(DefaultLayoutTextBoxType.TEXT, question.getValue() ?: ""))
        return result
    }

    fun createNumeric(question: NumericQuestion): MutableList<DefaultLayoutItem> {
        val result = buildCommon(question)
        result.add(
            DefaultLayoutTextBox(
                DefaultLayoutTextBoxType.NUMERIC, question.getValue()
                    ?: ""
            )
        )
        return result
    }

    fun createNotSupported(): MutableList<DefaultLayoutItem> {
        val result = mutableListOf<DefaultLayoutItem>()
        result.add(DefaultLayoutIconLabel(QuestionText("Question is not supported", false), R.drawable.ic_info_black))
        return result
    }

    fun createSingle(question: SingleQuestion): MutableList<DefaultLayoutItem> {
        val result = buildCommon(question)

        var selectedCode = ""
        question.selected()?.let {
            selectedCode = it.code
        }

        question.answers.forEach {
            if (it.isHeader) {
                if (!it.text.isEmpty()) {
                    result.add(DefaultLayoutLabel(DefaultLabelType.TEXT, it.text, 1))
                }
            } else {
                result.add(DefaultLayoutClickLabel(DefaultLayoutClickLabelType.RADIO, selectedCode == it.code, 1, it))
            }

            // Nested answer always not group header
            it.answers.forEach { nestedAnswer ->
                result.add(DefaultLayoutClickLabel(DefaultLayoutClickLabelType.RADIO, selectedCode == nestedAnswer.code, 2, nestedAnswer))
            }
        }

        return result
    }

    fun createMulti(question: MultiQuestion): MutableList<DefaultLayoutItem> {
        val result = buildCommon(question)

        question.answers.forEach {
            if (it.isHeader) {
                if (!it.text.isEmpty()) {
                    result.add(DefaultLayoutLabel(DefaultLabelType.TEXT, it.text, 1))
                }
            } else {
                result.add(DefaultLayoutClickLabel(DefaultLayoutClickLabelType.CHECKBOX, question.get(it), 1, it))
            }

            it.answers.forEach { nestedAnswer ->
                result.add(DefaultLayoutClickLabel(DefaultLayoutClickLabelType.CHECKBOX, question.get(it), 2, nestedAnswer))
            }
        }

        return result
    }
}